contracts/lens/Quoter.sol

```diff
@@ -0,0 +1,396 @@
+        }
+    }
+
+    function quoteExactOutput(ExactOutputParams memory params)
+        external
+        returns (
+            int128[] memory deltaAmounts,
+            uint160[] memory sqrtPriceX96AfterList,
+            uint32[] memory initializedTicksLoadedList
+        )
+    {
+        try manager.lock(abi.encode(SwapInfo(SwapType.ExactOutput, abi.encode(params)))) {}
+        catch (bytes memory reason) {
+            return _handleRevert(reason);
+        }
+    }
+
+    function quoteExactOutputBatch(ExactOutputSingleBatchParams memory params)
+        external
+        returns (
+            IQuoter.PoolDeltas[] memory deltas,
+            uint160[] memory sqrtPriceX96AfterList,
+            uint32[] memory initializedTicksLoadedList
+        )
+    {
+        if (
+            params.zeroForOnes.length != params.recipients.length
+                || params.recipients.length != params.amountOuts.length
+                || params.amountOuts.length != params.sqrtPriceLimitX96s.length
+                || params.sqrtPriceLimitX96s.length != params.hookData.length
+        ) {
+            revert InvalidQuoteBatchParams();
+        }
+
+        deltas = new IQuoter.PoolDeltas[](params.amountOuts.length);
+        sqrtPriceX96AfterList = new uint160[](params.amountOuts.length);
+        initializedTicksLoadedList = new uint32[](params.amountOuts.length);
+
+        for (uint256 i = 0; i < params.amountOuts.length; i++) {
+            ExactOutputSingleParams memory singleParams = ExactOutputSingleParams({
+                poolKey: params.poolKey,
+                zeroForOne: params.zeroForOnes[i],
+                recipient: params.recipients[i],
+                amountOut: params.amountOuts[i],
+                sqrtPriceLimitX96: params.sqrtPriceLimitX96s[i],
+                hookData: params.hookData[i]
+            });
+            (int128[] memory deltaAmounts, uint160 sqrtPriceX96After, uint32 initializedTicksLoaded) =
+                quoteExactOutputSingle(singleParams);
+
+            deltas[i] = IQuoter.PoolDeltas({currency0Delta: deltaAmounts[0], currency1Delta: deltaAmounts[1]});
+            sqrtPriceX96AfterList[i] = sqrtPriceX96After;
+            initializedTicksLoadedList[i] = initializedTicksLoaded;
+        }
+    }
+
+    function lockAcquired(bytes calldata encodedSwapIntention) external returns (bytes memory) {
+        if (msg.sender != address(manager)) {
+            revert InvalidLockAcquiredSender();
+        }
+
+        SwapInfo memory swapInfo = abi.decode(encodedSwapIntention, (SwapInfo));
+
+        if (swapInfo.swapType == SwapType.ExactInputSingle) {
+            (BalanceDelta deltas, uint160 sqrtPriceX96After, int24 tickAfter) =
+                _quoteExactInputSingle(abi.decode(swapInfo.params, (ExactInputSingleParams)));
+
+            bytes memory result = abi.encode(deltas, sqrtPriceX96After, tickAfter);
+            assembly {
+                revert(add(0x20, result), mload(result))
+            }
```

Comment:
this branch is repeated over and over, can factor out `result` and revert at the end?